opponentplay

lemz1 opened the pull request:
Good afternoon,
I need to repost this pull request (#2770), since I was kinda stupid, but it isn't just a repost.
I updated it to the new code and changed some stuff to have less variables and less lines.

### What does this do ###
This adds the gameplay option 'opponentplay'. If turned on lets you play the opponent chart.
It isn't a big feature but yeah.

lemz1 committed:
opponentplay

tposejank commented:
sick

HiroMizuki commented:
sick

CrusherMods commented:
sick

MayoALT commented:
sick

SrtHero278 commented:
sick

MasterDirector99 commented:
sick

JorgeMiguel22 commented:
why it doesnt appear anything about opponent play?

HiroMizuki commented:
> why it doesnt appear anything about opponent play?

what

JorgeMiguel22 commented:
i searched in the options and didnt find the opponent play option

HiroMizuki commented:
> i searched in the options and didnt find the opponent play option

are u sure that u coded everything changed at the PR?
or are u sure that u compiled the right source?

JorgeMiguel22 commented:
yeah i changed all the files needed

SrtHero278 commented:
If you're looking through the settings through the main menu. That's not how you find it. You gotta select a song and press `control`.

JorgeMiguel22 commented:
oh it worked, thx for helping

Megalonumber0ne commented:
I coded everything in correctly but when i went to play a song (I'll use Dadbattle for an example), all of the notes appear on the opponent's side, including BF's, also when BF and DD sync both of the notes appear at the same time and can only be pressed by either the AI or you (It's mostly the AI that gets the notes)
![image](https://user-images.githubusercontent.com/87905744/149919922-7b6273bc-83b3-4acc-9ecc-a37db8261f1e.png)
![image](https://user-images.githubusercontent.com/87905744/149919978-1ee8c414-d397-4a4f-8beb-f45e08bb2526.png)

lemz1 commented:
> I coded everything in correctly but when i went to play a song (I'll use Dadbattle for an example), all of the notes appear on the opponent's side, including BF's, also when BF and DD sync both of the notes appear at the same time and can only be pressed by either the AI or you (It's mostly the AI that gets the notes) ![image](https://user-images.githubusercontent.com/87905744/149919922-7b6273bc-83b3-4acc-9ecc-a37db8261f1e.png) ![image](https://user-images.githubusercontent.com/87905744/149919978-1ee8c414-d397-4a4f-8beb-f45e08bb2526.png)
> 
> EDIT: I have a screenshot of the notes mixing ![image](https://user-images.githubusercontent.com/87905744/149920395-a88b8e93-770a-4c15-a6c1-4b74b1663633.png)

The note generation is somewhere in line 1870.
Are you sure that you put in
```
if (songNotes[1] <= 3 && opponentChart)
{
	gottaHitNote = !section.mustHitSection;
}
```
Maybe you did a wrong symbol instead of `<=`

Megalonumber0ne commented:
Even after checking all of the code that was given to me and re-building the game i still got the same result

Megalonumber0ne commented:
Uploading Vide1.mp4…

Gui-iago commented:
nice! merged this into my fork and it works perfectly, I've seen some bugs of course (some of which I managed to fix) but here are all of the bugs I could find

* since Dad is now BF when Opponent Mode is on, Lua Scripts are completely swapped and can present issues (such as animations not working properly if lua scripts are present, or noteskins being swapped, etc)
* Opponent Mode Dad can't hold notes, his animation stops looping like it should if he was actually Player 2 if you try to do so, and it loops BF's animation instead
* GF Notes make it so Dad **and** GF sing together
  Solution: on line 4013, add "&& !gfNote" in "if(opponentChart) {"
  
* Alt Anims and Hey Notes are completely bugged, or straight up don't work sometimes (try playing Cocoa or Eggnog)
* sometimes (rarely) the charts don't swap, nothing major
* Miss Animations won't play if Ghost Tapping is disabled for Player 2 (try any Week 3 song with Ghost Tapping off)

that everything I know!

XtraXD1 commented:
sick

lemz3 committed:
Update PlayState.hx

Gui-iago commented:
(latest update)


https://user-images.githubusercontent.com/45212377/150021456-a5daa0f5-11f9-4929-a9bd-0b3145fc5f4c.mp4


https://user-images.githubusercontent.com/45212377/150021880-62782d0d-c79b-424d-807b-e5dc6835fbec.mp4

the same doesn't happen for BF

lemz1 committed:
bruh

lemz1 commented:
> (latest update)
> 2022-01-18.17-27-08.mp4
> 2022-01-18.17-30-26.mp4
> 
> the same doesn't happen for BF

Are you sure you coded everything right. The thing with bf dying doesn't really make sense, since I didn't change how health works.

Gui-iago commented:
> > (latest update)
> > 2022-01-18.17-27-08.mp4
> > 2022-01-18.17-30-26.mp4
> > the same doesn't happen for BF
> 
> Are you sure you coded everything right. The thing with bf dying doesn't really make sense, since I didn't change how health works. Also I noticed that you are using "Bedrock Engine". There's probably different code

the latest commit fixed it

Stilic commented:
> Also I noticed that you are using "Bedrock Engine".

It is his own fork of Pysch Engine with custom features from different PRs.

Gui-iago commented:
BE doesn't change the code as much, it's just that I probably didn't merge the changes correctly, yet it worked with the latest commit fine

lemz1 committed:
fixes crash when hitting note before curStep 0

crowplexus reviewed (approved):
epic

stilic reviewed (approved):
so epic than i will shit from fear if @ShadowMario or @bbpanzu doesn't merge this pr.

lemz1 committed:
Merge branch 'ShadowMario:main' into opponentplay

lemz1 committed:
yay animation works now

lemz1 committed:
miss anims work

Gui-iago commented:
![image](https://user-images.githubusercontent.com/45212377/152656374-8e3d169c-8254-4462-b0a2-76a75f82ab70.png)
this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error

lemz1 commented:
> ![image](https://user-images.githubusercontent.com/45212377/152656374-8e3d169c-8254-4462-b0a2-76a75f82ab70.png) this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error

Does it tell you which file is the problem? If yes then I can probably help else it will be harder

Gui-iago commented:
> > ![image](https://user-images.githubusercontent.com/45212377/152656374-8e3d169c-8254-4462-b0a2-76a75f82ab70.png) this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error
> 
> Does it tell you which file is the problem? If yes then I can probably help else it will be harder

no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off

Gui-iago commented:
just checked, also happens if botplay is on

lemz1 commented:
> 

> > > ![image](https://user-images.githubusercontent.com/45212377/152656374-8e3d169c-8254-4462-b0a2-76a75f82ab70.png) this started happening after the latest two commits, while I think they worked, once you hit a note (not even on curStep 0), it crashes and gives this error
> > 
> > 
> > Does it tell you which file is the problem? If yes then I can probably help else it will be harder
> 
> no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off

Can you give me your code like upload it to drive or if you have it in github can you tell me which repository.
I can then thoroughly look at it

Gui-iago commented:
> > no, but the error is always the same no matter which song I choose, it doesn't happen if I don't hit any notes, and it happens even if opponentplay is turned off
> 
> Can you give me your code like upload it to drive or if you have it in github can you tell me which repository. I can then thoroughly look at it
https://github.com/Gui-iago/FNF-BedrockEngine/commit/b178ebbfa5a3dbe658da2589440fa5538d2c36ce


https://user-images.githubusercontent.com/45212377/152657238-21654b8e-b80b-4f58-aa83-355b97928b82.mp4

the fixes probably work though, as bf's loops are shorter

lemz1 commented:
I think you need to remove in line 4388 
**_{_**

Gui-iago commented:
figured it out, it was an issue with Conductor and json shit, rather than your fork

keep up the good work!

lemz1 commented:
> figured it out, it was an issue with Conductor and json shit, rather than your fork

nice, I am glad you figured it out, because I was being too stupid to figure it out

> keep up the good work!

I'll try

lemz1 committed:
Merge branch 'ShadowMario:main' into opponentplay

lemz1 committed:
might be a fix for character specific events

lemz1 committed:
Merge branch 'opponentplay' of https://github.com/lemz1/FNF-PsychEngine into opponentplay

lemz1 commented:
[This](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a) needs to be turned of depending of what goodNoteHit does.
For example it needs to be turned off if you use something like,
```
function goodNoteHit()

	accuracyUpdateShitOrSomethingLikeThat()

end
```
since the accuracy update needs to be done everytime the note is hit.
It doesn't need to be turned off if you use something like,
```
function goodNoteHit()

	debugPrint('bf sings')

end
```
since it is supposed to activate when bf hits a note.

lemz1 commented:
Please tell me if there are any bugs or crashes because of [this](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a)

tposejank commented:
> [This](https://github.com/ShadowMario/FNF-PsychEngine/pull/3549/commits/4a40edbb2b30512cd2f67cd888622e6b2678d40a) needs to be turned off depending on what goodNoteHit or opponentNoteHit does. For example it needs to be turned off if you use something like,
> 
> ```
> function goodNoteHit()
> 
> 	accuracyUpdateShitOrSomethingLikeThat()
> 
> end
> ```
> 
> since the accuracy update needs to be done everytime the note is hit. It can be turned off by doing this
> 
> ```
> function onCreate()
> 
>     setProperty('noteHitFix', false)
> 
> end
> ```
> 
> It doesn't need to be turned off if you use something like,
> 
> ```
> function goodNoteHit()
> 
> 	debugPrint('bf sings')
> 
> end
> ```
> 
> since it is supposed to activate when bf hits a note.

https://c.tenor.com/DuThn51FjPcAAAAC/nerd-emoji-nerd.gif

frantastic24 commented:
having botplay and opponent chart turned on at the same time seems to make the opponent not reset back to idle after singing

lemz1 committed:
should be fixed now

lemz1 committed:
Merge branch 'ShadowMario:main' into opponentplay

lemz1 committed:
crash doesn't happen anymore when opening 'adjust delay and combo'

lemz1 committed:
now please be actually fixed

Gui-iago commented:
there's a glitch with the Cam Zoom event, but I found the solution pretty quickly

![image](https://user-images.githubusercontent.com/45212377/155579113-0f04f1fe-a541-4b21-8ea1-a0f3552bdd8a.png)

lemz1 committed:
fix for Cam Zoom Event by Gui-iago

lemz1 committed:
opponentplay is false on game start

also when you go back to the titlescreen

lemz1 committed:
botPlay anims work

why do I not pay more attention.
Fixing the crash resulted in botplay animation lasting forever

lemz1 committed:
IQ below average

lemz1 committed:
longer bf animation when in opponentplay

FletcherMC0329 commented:
Nice but the only problem is lua swapped when opponentplay.

lemz1 committed:
improved notehitFix

You can now use notehitfix and normal notehit while using multiple scripts.
Example Whitty Definitive Edition:
the accuracy function is just goodNoteHit since it is only supposed to happen when you hit it.
ballistic shake when whitty sings uses opponentNoteHitFix

FletcherMC0329 commented:
Ok

lemz1 committed:
grammar mistake, I lose

Gui-iago commented on source/GameplayChangersSubstate.hx:
```diff
@@ -91,7 +91,7 @@ class GameplayChangersSubstate extends MusicBeatSubstate
 		optionsArray.push(option);
 
-		var option:GameplayOption = new GameplayOption('Play Opponents Chart', 'opponentplay', 'bool', false);
+		var option:GameplayOption = new GameplayOption("Play Opponent's Chart", 'opponentplay', 'bool', false);
```
I think a better name for this could be "Play as Opponent"

lemz1 replied:
shish kebab

lemz1 committed:
Update GameplayChangersSubstate.hx

lemz1 committed:
animations work properly

frantastic24 commented:
this seems to break gf characters singing when the option is on, they don't return to their idle

https://user-images.githubusercontent.com/95725102/156490750-f4516eef-ebef-4801-822f-e83014d32899.mp4

lemz1 committed:
anims should be fixed now

lemz1 committed:
this is better

you won't be able to make gf to the sing anim forever but this will be less code so I think that this is for the better

lemz1 committed:
Merge branch 'main' into opponentplay

lemz1 committed:
Update PlayState.hx

lemz1 committed:
Merge branch 'main' into opponentplay

BeastlyGhost commented:
oh yeah since this is closed and finished now and won't be merged, I recommend to everyone who wants this feature to simply merge them to your own forks or private builds
